Better frameback

gigabo opened the pull request:
This is a big change to how frameback (and the history stack generally) is managed, but the changes in behavior are pretty straightforward.

Previously:
- Frameback
    - Each click into a frame destroyed and recreated the frame.
    - Any click _within_ the frame caused a full browser navigation.
- Generally
    - Client transition options (`reuseDom`, `bundleData`) only applied on first click, not on forward/backward browser nav.

With this changeset:
- Frameback
    - A click into a frame _may_ perform a client transition _within_ the frame to the new page.
        - Client transition options (`reuseDom`, `bundleData`) are available.
    - A click _within_ the frame may perform a client transition.
- Generally
    - Client transition options work on forward/backward browser nav.

Sounds simple, right? :feelsgood: 

Some things actually _did_ get simpler.  Some things will make your head hurt.  I've tried to address those with commentary.  Please don't be shy about calling out confusing bits.

gigabo commented on packages/react-server-test-pages/pages/navigation/playground.js:
```diff
@@ -33,27 +33,58 @@ const PagePointer = ({page, row}) => <div className="page-pointer">
+const FBCT     = (props) => <FBL {...props} link={{reuseFrame:true}}>CT</FBL>
+const FBCTBD   = (props) => <FBL {...props} link={{reuseFrame:true, bundleData:true}}>CT/BD</FBL>
+const FBCTRD   = (props) => <FBL {...props} link={{reuseFrame:true, reuseDom:true}}>CT/RD</FBL>
+const FBCTBDRD = (props) => <FBL {...props} link={{reuseFrame:true, bundleData:true, reuseDom:true}}>CT/BD/RD</FBL>
```
Have a lot of links now.  Gave them shorter names.  Added a legend to the page.

![image](https://cloud.githubusercontent.com/assets/115908/14577463/3d16fd36-0329-11e6-948c-6d7c77d43696.png)

gigabo commented on packages/react-server/core/ClientController.js:
```diff
@@ -69,7 +76,6 @@ class ClientController extends EventEmitter {
 		ReactServerAgent.cache().rehydrate(dehydratedState.InitialContext['ReactServerAgent.cache']);
 		this.mountNode = document.getElementById('content');
 
-		this._setupFramebackController();
```
This is attached to the request context now, since the navigator needs to get at it.

gigabo commented on packages/react-server/core/ClientController.js:
```diff
@@ -172,27 +243,6 @@ class ClientController extends EventEmitter {
-					reactServerFrame: pushFrame(),
-				}, null, path);
-			}
-
```
We were pushing state after running `handleRoute`.  Moved it up into `_startRequest`.

gigabo commented on packages/react-server/core/ClientController.js:
```diff
@@ -558,7 +608,7 @@ class ClientController extends EventEmitter {
 					// first unmounting from React and then
 					// destroying the DOM node.
-					React.unmountComponentAtNode(root);
+					ReactDOM.unmountComponentAtNode(root);
```
Oops, this is just an incidental fix to hush React's deprecation warning.

roblg replied:
Incidental or not, good change :)

roblg commented:
I tried to pull this branch to play with it, and I can't get `react-server-cli` to work in `react-server-test-pages`.

I ran:
```
lerna bootstrap
cd packages/react-server-test-pages
npm start
```

and got:

`sh: react-server-cli: command not found`

I can't run `react-server-cli` via `node_modules` because `lerna bootstrap` does the funky not-an-npm-link thing.

I also ran into #98 futzing around with things.

Any advice on how I might go about getting the server running? :grin:

roblg commented:
Figured out a workaround: used `npm install ../react-server-cli` from the `react-server-test-pages` package and all was well.

roblg commented on packages/react-server/core/ClientController.js:
```diff
@@ -151,10 +156,76 @@ class ClientController extends EventEmitter {
+		// If this is a History.events.PUSHSTATE navigation,
+		// and we have control of the navigation bar (we're
+		// not in a frameback frame) we should change the URL
+		// in the bar location bar before rendering.
```
typo: "in the bar location bar"

gigabo replied:
Heh, wonder how long [that's been there](https://github.com/redfin/react-server/pull/97/files#diff-db9055542b789c676642bfe7f729f1ecL175).  Will fix.

gigabo commented:
> used npm install ../react-server-cli from the react-server-test-pages package and all was well.

Ugh.  Yeah, I'm not sure if we're somehow mis-using `lerna` or if it just doesn't do what we want it to do: Link together the cross dependencies within our monorepo.

The `react-server-test-pages` in this branch also needs the `react-server` from this branch.

roblg commented:
`react-server` worked, I think. I think the issue that I was running into w/ `react-server-cli` is that is has a binary defined in package.json.

roblg commented:
BTW, I reviewed it I think as throughly as I'll be able to. I pulled it down and it works, and it seems to remember things through forward and back buttons. 👍

gigabo commented:
Awesome.  Thanks for looking at it.

BTW - I think I got one behavior _wrong_ here that I'm going to address before I merge: Frameback links in a frame are only safe to convert to normal client transitions _if `reuseFrame` is set_.  Currently they're _always_ converted.  If `reuseFrame` isn't set, I think it's (unfortunately) only safe to convert them to top-level full browser navigation.

gigabo commented on packages/react-server/core/ClientController.js:
```diff
@@ -169,30 +240,9 @@ class ClientController extends EventEmitter {
-			} else if (type === History.events.PAGELOAD && this._history && this._history.canClientNavigate()) {
-				this._history.replaceState({
-					reactServerFrame: pushFrame(),
-				}, null, path);
```
Oh, just realized I introduced a bug here.  It's tough to see because this whole block of code had to _move_, but I also eliminated the `replaceState` on `PAGELOAD`.  My reasoning was that any client transition away from the initial page would handle the `replaceState`, so the one at load time seemed redundant.  I had forgotten about non-`react-server` users of the history navigation API. :hurtrealbad:

gigabo commented:
> If reuseFrame isn't set, I think it's (unfortunately) only safe to convert them to top-level full browser navigation.

Turned out not be as bad as all that.  Converting instead to a direct non-`reuseFrame` frameback navigation.  Nice that things like this can finally be [pretty straightforward](https://github.com/redfin/react-server/pull/97/commits/cc524d34e72d61e6087474671be8fbce0eaa7c2e).  Used to be that any change to frameback logic was a nightmare. 😱

gigabo commented:
(Moving this from edited PR description.  Progress bar didn't how up in list view... maybe it will if this is in a new comment?)

Still have a few of items to deal with that came up since this PR was created:

- [x] Handle non-`reuseFrame` frameback links within the frame.
- [ ] Set history nav state on page load
- [ ] Test interaction with non-`react-server` history nav users

gigabo commented:
Okay, I've addressed the remaining issues here.

Going to merge once the build passes.  If you want to look at this before I do, stop me before then. 😛
